File: embedded/unittest/platform_test/pubsub_test.rs
//! Unit tests for the platform-level bounded Pub/Sub ring buffer.

use std::{sync::Arc, time::Duration};

use super::*;

fn message<T>(status: ReceiveStatus<T>) -> Arc<T> {
    match status {
        ReceiveStatus::Message(value) => value,
        ReceiveStatus::Timeout => panic!("expected a message, received timeout"),
        ReceiveStatus::Closed => panic!("expected a message, topic was closed"),
    }
}

/// Every subscriber receives the same immutable shared allocation.
#[test]
fn broadcasts_one_shared_message_to_all_subscribers() {
    let topic = Topic::new("test/topic", 2).unwrap();
    let first = topic.subscribe("first").unwrap();
    let second = topic.subscribe("second").unwrap();
    let publisher = topic.publisher().unwrap();

    assert_eq!(publisher.publish(42_u32).unwrap(), 2);
    let first_message = message(first.receive().unwrap());
    let second_message = message(second.receive().unwrap());

    assert_eq!(*first_message, 42);
    assert!(Arc::ptr_eq(&first_message, &second_message));
}

/// New subscribers intentionally start with the next publication.
#[test]
fn new_subscriber_skips_retained_history() {
    let topic = Topic::new("test/topic", 2).unwrap();
    let publisher = topic.publisher().unwrap();
    publisher.publish(1_u32).unwrap();
    let subscriber = topic.subscribe("late-reader").unwrap();
    assert!(matches!(
        subscriber.try_receive().unwrap(),
        ReceiveStatus::Timeout
    ));
    publisher.publish(2_u32).unwrap();
    assert_eq!(*message(subscriber.receive().unwrap()), 2);
}

/// A slow subscriber resumes from the oldest retained message and counts loss.
#[test]
fn overwrites_oldest_data_and_counts_drops() {
    let topic = Topic::new("test/topic", 2).unwrap();
    let subscriber = topic.subscribe("slow-reader").unwrap();
    let publisher = topic.publisher().unwrap();
    publisher.publish(1_u32).unwrap();
    publisher.publish(2_u32).unwrap();
    publisher.publish(3_u32).unwrap();

    assert_eq!(*message(subscriber.receive().unwrap()), 2);
    assert_eq!(*message(subscriber.receive().unwrap()), 3);
    assert_eq!(subscriber.dropped_messages(), 1);
}

/// Subscriber cursors advance independently through the same ring.
#[test]
fn maintains_independent_subscriber_cursors() {
    let topic = Topic::new("test/topic", 3).unwrap();
    let fast = topic.subscribe("fast").unwrap();
    let slow = topic.subscribe("slow").unwrap();
    let publisher = topic.publisher().unwrap();
    publisher.publish(10_u32).unwrap();
    publisher.publish(20_u32).unwrap();

    assert_eq!(*message(fast.receive().unwrap()), 10);
    assert_eq!(*message(fast.receive().unwrap()), 20);
    assert_eq!(*message(slow.receive().unwrap()), 10);
    assert_eq!(*message(slow.receive().unwrap()), 20);
}

/// One worker wait set can be raised by any of its subscribed topics.
#[test]
fn wait_set_supports_multiple_topics() {
    let first_topic = Topic::<u32>::new("first", 2).unwrap();
    let second_topic = Topic::<u32>::new("second", 2).unwrap();
    let wait_set = Arc::new(TopicWaitSet::default());
    let _first = first_topic
        .subscribe_with_wait_set("worker", Arc::clone(&wait_set), 0)
        .unwrap();
    let second = second_topic
        .subscribe_with_wait_set("worker", Arc::clone(&wait_set), 1)
        .unwrap();
    let publisher = second_topic.publisher().unwrap();
    publisher.publish(7_u32).unwrap();

    assert_eq!(
        wait_set
            .wait_timeout(0b11, Duration::from_millis(50))
            .unwrap(),
        0b10
    );
    assert_eq!(*message(second.receive().unwrap()), 7);
    assert_eq!(wait_set.ready_topics(), 0);
}

/// Subscribers wake and report closure when the final publisher exits.
#[test]
fn closes_subscribers_after_last_publisher_exits() {
    let topic = Topic::<u32>::new("test/topic", 1).unwrap();
    let subscriber = topic.subscribe("reader").unwrap();
    let publisher = topic.publisher().unwrap();
    drop(publisher);
    assert!(matches!(
        subscriber.receive().unwrap(),
        ReceiveStatus::Closed
    ));
}

#[test]
fn detects_sequence_wrap_without_ordering_errors() {
    let last = TopicSequence {
        epoch: 3,
        value: u64::MAX,
    };
    assert_eq!(next_sequence(last), TopicSequence { epoch: 4, value: 0 });
}

#[test]
fn rejects_zero_capacity_and_invalid_wait_bit() {
    assert!(Topic::<u32>::new("test/topic", 0).is_err());
    assert!(TopicWaitSet::bit(64).is_err());
}

File: embedded_C/unittest/platform_test/pubsub_test.cpp
#include <chrono>
#include <cstdint>
#include <limits>
#include <memory>

#include "1_Platform/pubsub/pubsub.hpp"
#include "unittest/test.hpp"

VISTA_TEST(pubsub_broadcasts_one_shared_message) {
    vista::platform::Topic<int> topic("test/topic", 2);
    auto first = topic.subscribe("first");
    auto second = topic.subscribe("second");
    auto publisher = topic.publisher();

    VISTA_CHECK(publisher.publish(42) == 2);
    std::shared_ptr<const int> first_message;
    std::shared_ptr<const int> second_message;
    VISTA_CHECK(first.receive(first_message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(second.receive(second_message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*first_message == 42);
    VISTA_CHECK(first_message == second_message);
}

VISTA_TEST(pubsub_closes_after_last_publisher) {
    vista::platform::Topic<int> topic("test/topic", 1);
    auto subscriber = topic.subscribe("reader");
    {
        auto publisher = topic.publisher();
    }
    std::shared_ptr<const int> message;
    VISTA_CHECK(subscriber.receive(message) == vista::platform::ReceiveStatus::closed);
}

VISTA_TEST(pubsub_rejects_zero_capacity) {
    VISTA_CHECK_THROWS(vista::platform::Topic<int>("bad", 0));
}

VISTA_TEST(pubsub_new_subscriber_starts_at_next_published_message) {
    vista::platform::Topic<int> topic("test/latest", 4);
    auto publisher = topic.publisher();
    VISTA_CHECK(publisher.publish(1) == 0);

    auto subscriber = topic.subscribe("late-reader");
    VISTA_CHECK(publisher.publish(2) == 1);

    std::shared_ptr<const int> message;
    VISTA_CHECK(
        subscriber.receive(message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*message == 2);
}

VISTA_TEST(pubsub_ring_overwrites_oldest_and_counts_subscriber_drops) {
    vista::platform::Topic<int> topic("test/ring", 2);
    auto subscriber = topic.subscribe("slow-reader");
    auto publisher = topic.publisher();

    publisher.publish(1);
    publisher.publish(2);
    publisher.publish(3);

    std::shared_ptr<const int> message;
    VISTA_CHECK(
        subscriber.receive(message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*message == 2);
    VISTA_CHECK(subscriber.dropped_messages() == 1);
    VISTA_CHECK(
        subscriber.receive(message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*message == 3);
}

VISTA_TEST(pubsub_subscribers_advance_independent_cursors) {
    vista::platform::Topic<int> topic("test/independent", 3);
    auto fast = topic.subscribe("fast");
    auto slow = topic.subscribe("slow");
    auto publisher = topic.publisher();
    publisher.publish(10);
    publisher.publish(20);

    std::shared_ptr<const int> fast_message;
    std::shared_ptr<const int> slow_message;
    VISTA_CHECK(fast.receive(fast_message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(fast.receive(fast_message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*fast_message == 20);

    VISTA_CHECK(slow.receive(slow_message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*slow_message == 10);
    VISTA_CHECK(slow.dropped_messages() == 0);
}

VISTA_TEST(pubsub_wait_set_reports_any_of_multiple_topics) {
    auto wait_set = std::make_shared<vista::platform::TopicWaitSet>();
    vista::platform::Topic<int> first_topic("test/first", 2);
    vista::platform::Topic<int> second_topic("test/second", 2);
    auto first = first_topic.subscribe("worker-first", wait_set, 0);
    auto second = second_topic.subscribe("worker-second", wait_set, 1);
    auto first_publisher = first_topic.publisher();
    auto second_publisher = second_topic.publisher();

    second_publisher.publish(22);
    const auto ready = wait_set->wait_for(
        vista::platform::TopicWaitSet::all_topics,
        std::chrono::milliseconds(10));
    VISTA_CHECK((ready & vista::platform::TopicWaitSet::bit(0)) == 0);
    VISTA_CHECK((ready & vista::platform::TopicWaitSet::bit(1)) != 0);

    std::shared_ptr<const int> message;
    VISTA_CHECK(second.receive(message) == vista::platform::ReceiveStatus::message);
    VISTA_CHECK(*message == 22);
    VISTA_CHECK(
        (wait_set->ready_topics() & vista::platform::TopicWaitSet::bit(1)) == 0);
}

VISTA_TEST(pubsub_sequence_wraps_with_an_incremented_epoch) {
    const vista::platform::TopicSequence before{
        7, std::numeric_limits<std::uint64_t>::max()};
    const auto after = vista::platform::detail::next_topic_sequence(before);
    VISTA_CHECK(after.epoch == 8);
    VISTA_CHECK(after.value == 0);

    const vista::platform::TopicSequence cursor{
        7, std::numeric_limits<std::uint64_t>::max() - 2};
    const vista::platform::TopicSequence next{8, 3};
    VISTA_CHECK(
        vista::platform::detail::sequence_distance_saturated(cursor, next) == 6);
}

VISTA_TEST(pubsub_wait_set_rejects_more_than_64_topic_bits) {
    VISTA_CHECK_THROWS(vista::platform::TopicWaitSet::bit(64));
}
